### ISSUE TITLE
Race condition in suspended transactions across Coroutine Scopes

### ISSUE THREAD

**Issue by peterfigure** [JetBrains/Exposed#1183]
I've noticed this problem happening when two `newSuspendedTransactions` are competing on the same table from different CoroutineScopes, when `autoCommit` in the connection pool (Hikari) is set to `false`:

```
org.jetbrains.exposed.exceptions.ExposedSQLException: org.postgresql.util.PSQLException: Cannot change transaction isolation level in the middle of a transaction.
	at org.jetbrains.exposed.sql.statements.Statement.executeIn$exposed_core(Statement.kt:50)
	at org.jetbrains.exposed.sql.Transaction.exec(Transaction.kt:129)
	at org.jetbrains.exposed.sql.Transaction.exec(Transaction.kt:115)
	at org.jetbrains.exposed.sql.statements.Statement.execute(Statement.kt:29)
	at org.jetbrains.exposed.sql.QueriesKt.insert(Queries.kt:70)
```

Simple illustration project: https://github.com/peterfigure/exposed-suspended-transactions

It looks like this happens when `db.connector()` returns a previously unused connection... but why would Postgres think that a transaction on this new connection is in progress?

**Comment by peterfigure** [comment]
@Tapac any thoughts? it seems pretty unfortunate not able to turn off autocommit

**Comment by Tapac** [comment]
Looks like the problem here is that you define an explicit scheme on HikaryConfig and that initializes a transaction with default isolation.
If you comment out `//    config.schema = "test"` the test will succeed.

**Comment by peterfigure** [comment]
Thanks @Tapac - that's a bit unexpected, is it a function of Hikari or Exposed to initialize a transaction on every connection when specifying the default schema?

**Comment by sickfar** [comment]
I have a strange behavior during my JUnit tests with Micronaut + Hikari
```
16:51:38.363 [main] INFO  com.zaxxer.hikari.HikariDataSource - HikariPool-7 - Shutdown initiated...
16:51:38.667 [main] INFO  com.zaxxer.hikari.HikariDataSource - HikariPool-7 - Shutdown completed.
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.422
...Some tests were run next
16:51:47.536 [main] INFO  com.zaxxer.hikari.HikariDataSource - HikariPool-10 - Start completed.
16:51:48.554 [DefaultDispatcher-worker-2 @coroutine#104] WARN  Exposed - Transaction attempt #1 failed: java.sql.SQLException: HikariDataSource HikariDataSource (HikariPool-7) has been closed.. Statement(s): *private*
org.jetbrains.exposed.exceptions.ExposedSQLException: java.sql.SQLException: HikariDataSource HikariDataSource (HikariPool-7) has been closed.
	at org.jetbrains.exposed.sql.statements.Statement.executeIn$exposed_core(Statement.kt:50)
	at org.jetbrains.exposed.sql.Transaction.exec(Transaction.kt:129)
	at org.jetbrains.exposed.sql.Query.count(Query.kt:177)
	at org.jetbrains.exposed.sql.Query.count(Query.kt:170)
	at org.jetbrains.exposed.sql.IterableExKt$mapLazy$1.count(IterableEx.kt:123)
```
Ok, tests utilizing HikariPool-7 have been finished some time ago, and I have `TransactionManager.closeAndUnregister(database)` on PreDestroy hook
So it looks like a leak between coroutines

It does not occur in production because I have only one pool in one JSM but it randomly fails my tests
Any ideas?

**Comment by Tapac** [comment]
@peterfigure ,as you can see from HikariCP [PoolBase](https://github.com/brettwooldridge/HikariCP/blob/f9b2c3d372b3f28911a872191ad62953c84bafa4/src/main/java/com/zaxxer/hikari/pool/PoolBase.java#L426) code it calls `connection.setSchema()`  function which leads to [PgConnection.setSchema()](https://github.com/pgjdbc/pgjdbc/blob/4595a5ae430ba5ee5463280d04c261d999813d0f/pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java#L1570) where the first query is executed and transaction (with value defined on a pool) is initialized.

**Comment by Tapac** [comment]
@sickfar, do you run the latest Exposed version? There was a fix in [0.30.1](https://github.com/JetBrains/Exposed/blob/master/docs/ChangeLog.md#0301) about count/empty memory leak.

**Comment by sickfar** [comment]
@Tapac, yes, that's exactly the version I did run it. It happen when I have 10 test JUnit test classes or more (each starting own data source). Switching to synchronous transaction solved the problem.

**Comment by Tapac** [comment]
@sickfar , it would be very helpful if you can provide a sample repo or just a whole test case (with some table mappings) to debug it. I found a bug in a suspended transactions but not sure that it's related to your problem.

**Comment by AlbRoehm** [comment]
Hey, i have encountered something kind of similar to this with the connection pool timeout. I already posted in slack https://kotlinlang.slack.com/archives/C0CG7E0A1/p1652110550581649 Should i add some more information here or do you think this is unrelated?

**Comment by bog-walk** [comment]
Fixed by #1943 and #1949
